# AfterWork Data Science: Data Wrangling with Python Project

## 1. Defining the Question

### a) Specifying the Data Analysis Question

Specify the reasearch question that you'll be answering.

> i.e. Which was the most improved country year 2017 - 2018 as per the GDP ranking?

1. Which bus companies had the highest breakdowns?
2. What were the top 3 reasons for bus delays?
3. How many students were on the buses when they broke down?
4. Which were the most frequent reasons for bus breakdowns?
5. What were the most frequent reasons for the bus running late?
6. What was the average delay time of each reason type?

### b) Defining the Metric for Success

How will you know that your solution will have satisfied your research question?

1. By being able to sort and see the bus companies with the highest breakdowns
2. By being able to tell the top 3 reasons for bus delays
3. By being able to tell the number of students that were on board the buses when they broke down
4. By being able to see which were the most frequent reasons for bus breakdowns from the data
5. By being able to see which were the most frequent reasons for bus running late from the data
6. By being able to tell the average delay time per reason of delay

### c) Understanding the context 

Provide some background information....

The Bus Breakdown and Delay system collects information from school bus vendors operating out in the field in real time. Bus staff that encounter delays during the route are instructed to radio the dispatcher at the bus vendor’s central office. The bus vendor staff are then instructed to log into the Bus Breakdown and Delay system to record the event and notify OPT. OPT customer service agents use this system to inform parents who call with questions regarding bus service. The Bus Breakdown and Delay system is publicly accessible and contains real time updates. All information in the system is entered by school bus vendor staff.

### d) Recording the Experimental Design

Describe the steps/approach that you will use to answer the given question.

1. Data cleaning
2. Data sorting 

### e) Data Relevance

How relevant was the provided data?

The data was sufficient and relevant to answer all the questions

## 2. Reading the Data

In [17]:
# Importing our libraries
# ---
#
import pandas as pd
import numpy as nm

In [18]:
# Load the data below
# --- 
# Dataset url = https://bit.ly/BusBreakdownDataset
# --- 
# 
Bus_breakdown = pd.read_csv('https://bit.ly/BusBreakdownDataset')
Bus_breakdown

School_Year  ...  School_Age_or_PreK
0        2015-2016  ...          School-Age
1        2015-2016  ...          School-Age
2        2015-2016  ...               Pre-K
3        2015-2016  ...          School-Age
4        2015-2016  ...          School-Age
...            ...  ...                 ...
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[281110 rows x 21 columns]

In [19]:
# Checking the first 5 rows of data
# ---
#
Bus_breakdown.head()

School_Year  Busbreakdown_ID  ... Breakdown_or_Running_Late School_Age_or_PreK
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

In [20]:
# Checking the last 5 rows of data
# ---
#
Bus_breakdown.tail()

School_Year  ...  School_Age_or_PreK
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[5 rows x 21 columns]

In [21]:
# Sample 10 rows of data
# ---
#
Bus_breakdown.sample(10)

School_Year  ...  School_Age_or_PreK
217286   2015-2016  ...          School-Age
9392     2015-2016  ...          School-Age
217392   2016-2017  ...          School-Age
117754   2017-2018  ...               Pre-K
37352    2016-2017  ...          School-Age
88380    2016-2017  ...          School-Age
118523   2018-2019  ...          School-Age
98309    2016-2017  ...          School-Age
109947   2017-2018  ...          School-Age
52632    2016-2017  ...          School-Age

[10 rows x 21 columns]

In [22]:
# Checking number of rows and columns
# ---
#  
Bus_breakdown.shape

(281110, 21)

In [23]:
# Checking datatypes
# ---
# 
Bus_breakdown.dtypes

School_Year                        object
Busbreakdown_ID                     int64
Run_Type                           object
Bus_No                             object
Route_Number                       object
Reason                             object
Schools_Serviced                   object
Occurred_On                        object
Created_On                         object
Boro                               object
Bus_Company_Name                   object
How_Long_Delayed                   object
Number_Of_Students_On_The_Bus       int64
Has_Contractor_Notified_Schools    object
Has_Contractor_Notified_Parents    object
Have_You_Alerted_OPT               object
Informed_On                        object
Incident_Number                    object
Last_Updated_On                    object
Breakdown_or_Running_Late          object
School_Age_or_PreK                 object
dtype: object

Record your observations below:

*   The dataset has 281110 rows and 21 columns.
*   The dataset has all columns with objects with the exception of Busbreakdown_ID and Number_Of_Students_On_The_Bus which have integer datatypes.



## 3. External Data Source Validation

You can make sure your data matches other credible information. This ensures that the measurements are roughly in line with what they should be and it serves as a check on what other things might be wrong in your dataset.

External validation can often be as simple as checking your data against some other data i.e. country population your dataset with country population on world bank data.

Replace the text found on this text cell with your external data source validation information.

## 4. Data Preparation

### Performing Data Cleaning

In [24]:
# Checking datatypes and missing entries of all the variables
# ---
# 
Bus_breakdown.dtypes

School_Year                        object
Busbreakdown_ID                     int64
Run_Type                           object
Bus_No                             object
Route_Number                       object
Reason                             object
Schools_Serviced                   object
Occurred_On                        object
Created_On                         object
Boro                               object
Bus_Company_Name                   object
How_Long_Delayed                   object
Number_Of_Students_On_The_Bus       int64
Has_Contractor_Notified_Schools    object
Has_Contractor_Notified_Parents    object
Have_You_Alerted_OPT               object
Informed_On                        object
Incident_Number                    object
Last_Updated_On                    object
Breakdown_or_Running_Late          object
School_Age_or_PreK                 object
dtype: object

In [25]:
Bus_breakdown.isnull()

School_Year  ...  School_Age_or_PreK
0             False  ...               False
1             False  ...               False
2             False  ...               False
3             False  ...               False
4             False  ...               False
...             ...  ...                 ...
281105        False  ...               False
281106        False  ...               False
281107        False  ...               False
281108        False  ...               False
281109        False  ...               False

[281110 rows x 21 columns]

In [26]:
Bus_breakdown.isnull().any()

School_Year                        False
Busbreakdown_ID                    False
Run_Type                            True
Bus_No                              True
Route_Number                        True
Reason                              True
Schools_Serviced                    True
Occurred_On                        False
Created_On                         False
Boro                                True
Bus_Company_Name                   False
How_Long_Delayed                    True
Number_Of_Students_On_The_Bus      False
Has_Contractor_Notified_Schools    False
Has_Contractor_Notified_Parents    False
Have_You_Alerted_OPT               False
Informed_On                        False
Incident_Number                     True
Last_Updated_On                    False
Breakdown_or_Running_Late          False
School_Age_or_PreK                 False
dtype: bool

In [27]:
Bus_breakdown.isnull().sum()

School_Year                             0
Busbreakdown_ID                         0
Run_Type                                3
Bus_No                                  9
Route_Number                            7
Reason                                  2
Schools_Serviced                        7
Occurred_On                             0
Created_On                              0
Boro                                13461
Bus_Company_Name                        0
How_Long_Delayed                    35608
Number_Of_Students_On_The_Bus           0
Has_Contractor_Notified_Schools         0
Has_Contractor_Notified_Parents         0
Have_You_Alerted_OPT                    0
Informed_On                             0
Incident_Number                    271627
Last_Updated_On                         0
Breakdown_or_Running_Late               0
School_Age_or_PreK                      0
dtype: int64

We observe the following from our dataset:

*   The dataset has no missing entries of all the variables
*   The dataset has Incident_number column with the most missing entries
* The dataset has 8 columns with missing entries 



In [28]:
# Standardizing your dataset i.e. variable renaming
# ---
# 
Bus_breakdown.columns = Bus_breakdown.columns.str.strip().str.upper().str.replace(' ', '_').str.replace('(', '').str.replace(')', '')
Bus_breakdown

SCHOOL_YEAR  ...  SCHOOL_AGE_OR_PREK
0        2015-2016  ...          School-Age
1        2015-2016  ...          School-Age
2        2015-2016  ...               Pre-K
3        2015-2016  ...          School-Age
4        2015-2016  ...          School-Age
...            ...  ...                 ...
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[281110 rows x 21 columns]

In [33]:
Bus_breakdown['HOW_LONG_DELAYED'] = Bus_breakdown['HOW_LONG_DELAYED'].str.strip().str.upper().str.replace(r"[a-zA-Z]",'').str.replace('[?,-,/,+,_,:,],\\,*,', '')  
Bus_breakdown = Bus_breakdown[Bus_breakdown['HOW_LONG_DELAYED'].notna()]
Bus_breakdown = Bus_breakdown.assign(HOW_LONG_DELAYED = lambda x: x['HOW_LONG_DELAYED'].str.extract('(\d+)'))
Bus_breakdown['HOW_LONG_DELAYED'].unique()
#Bus_breakdown['HOW_LONG_DELAYED'] = Bus_breakdown['HOW_LONG_DELAYED']

array(['20', '15', '30', '40', '10', '60', '35', '25', '5', '0', '45',
       '31', '46', '90', '1', '16', '1354', '1388', '1603', '61', '1425',
       '23', '105', '12', '8', '100', '55', '1403', '50', '13', '3', '7',
       '2', '401', '120', '6', '26', '9', '14', '4', '1310', '5819', '11',
       '800', '2640', '1427', '1376', '11658', '1420', '202340', '016224',
       '43', '65', '0634', '1489', '1003', '21', '115', '10481', '19',
       '302', '24', '17', '18', '156', '2557', '145', '52', '451', '22',
       '42', '59', '1450', '28', '10311', '27', '1375', '5698', '2359',
       '414', '0627', '454', '2085', '602', '9944', '1369', '1495',
       '1366', '08', '1390', '225', '197', '250', '300', '1435', '160',
       '2305', '2345', '1344', '104', '1457', '1494', '187', '1486',
       '1478', '2228', '1657', '146', '1254', '2015', '51', '1256',
       '2759', '445', '2254', '56', '34', '29', '2653', '202', '218',
       '75', '0933', '260', '190', '200', '54', '4104', '39', '882',

In [34]:
Bus_breakdown['HOW_LONG_DELAYED'] = Bus_breakdown['HOW_LONG_DELAYED'].astype(int)
Bus_breakdown

SCHOOL_YEAR  ...  SCHOOL_AGE_OR_PREK
1        2015-2016  ...          School-Age
2        2015-2016  ...               Pre-K
3        2015-2016  ...          School-Age
5        2015-2016  ...          School-Age
6        2015-2016  ...          School-Age
...            ...  ...                 ...
281104   2016-2017  ...               Pre-K
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[245002 rows x 21 columns]

We observe the following from our dataset:

*   The dataset now has clean data on the HOW_LONG_DELAYED column with intergers only
*   The dataset now has 245002 rows and 21 colums after cleaning of unusable data in HOW_LONG_DELAYED column



In [35]:
# Checking how many duplicate rows are there in the data
# ---
# 
Bus_breakdown[Bus_breakdown.duplicated()].shape

(0, 21)

We observe the following from our dataset:

*   There are no duplicate rows in the dataset



In [36]:
# Checking if any of the columns are all null
# ---
# 
Bus_breakdown.columns.isnull()

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False])

We observe the following from our dataset:

*   There are no columns which are all null



In [37]:
# Checking if any of the rows are all null
# ---
#
#Bus_breakdown.all(1).any()
Bus_breakdown.isna().any(axis=1)

1          True
2          True
3          True
5          True
6          True
          ...  
281104     True
281105     True
281106     True
281107    False
281109     True
Length: 245002, dtype: bool

We observe the following from our dataset:

*   There are no rows which are all null in the dataset



In [38]:
# Checking if the "Yes/No" fields contain only these 2 values
# for have_you_alerted_opt variable
# ---
# Hint: Use unique() function
#
Bus_breakdown_unique=  Bus_breakdown['HAVE_YOU_ALERTED_OPT'].unique()
list(Bus_breakdown_unique)

['No', 'Yes']

We observe the following from our dataset:

*   There are only yes and no values in the have_you_alerted_opt column



In [39]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_parents variable
# ---
# 
Bus_breakdown_unique=  Bus_breakdown['HAS_CONTRACTOR_NOTIFIED_PARENTS'].unique()
list(Bus_breakdown_unique)

['Yes', 'No']

We observe the following from our dataset:

*   There are only yes and no values in the has_contractor_notified_parents column



In [40]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_schools variable
# ---
#
Bus_breakdown_unique=  Bus_breakdown['HAS_CONTRACTOR_NOTIFIED_SCHOOLS'].unique()
list(Bus_breakdown_unique)  

['Yes', 'No']

We observe the following from our dataset:

*   There are only yes and no values in the has_contractor_notified_schools column



In [41]:
# Checking unique values in break_down_or_running_late variable to ensure there is no duplication
# ---
Bus_breakdown_late =  Bus_breakdown['BREAKDOWN_OR_RUNNING_LATE'].unique()
list(Bus_breakdown_late) 

['Running Late', 'Breakdown']

In [42]:
# Checking unique values in school_age_or_prek variable
# ---
# 
Bus_breakdown_late =  Bus_breakdown['SCHOOL_AGE_OR_PREK'].unique()
list(Bus_breakdown_late)

['School-Age', 'Pre-K']

In [43]:
# Checking unique values in school_year variable
# ---
#  
Bus_breakdown_late =  Bus_breakdown['SCHOOL_YEAR'].unique()
list(Bus_breakdown_late)

['2015-2016', '2018-2019', '2016-2017', '2017-2018']

In [44]:
# Checking unique values in reason variable
# ---
#  
Bus_breakdown_late =  Bus_breakdown['REASON'].unique()
list(Bus_breakdown_late)

['Heavy Traffic',
 'Other',
 'Mechanical Problem',
 'Flat Tire',
 'Problem Run',
 'Won`t Start',
 'Accident',
 'Late return from Field Trip',
 'Delayed by School',
 'Weather Conditions',
 nan]

In [45]:
# Checking unique values in run_type variable
# ---
#
Bus_breakdown_late =  Bus_breakdown['RUN_TYPE'].unique()
list(Bus_breakdown_late)  

['Special Ed AM Run',
 'Pre-K/EI',
 'General Ed AM Run',
 'General Ed Field Trip',
 'Special Ed PM Run',
 'General Ed PM Run',
 'Project Read PM Run',
 'Special Ed Field Trip',
 'Project Read AM Run',
 nan,
 'Project Read Field Trip']

In [46]:
# Checking unique values in boro variable
# ---
#
Bus_breakdown_late =  Bus_breakdown['BORO'].unique()
list(Bus_breakdown_late)   

['Manhattan',
 'Bronx',
 'Westchester',
 'Brooklyn',
 'Rockland County',
 'Nassau County',
 nan,
 'Queens',
 'Staten Island',
 'Connecticut',
 'New Jersey',
 'All Boroughs']

### Overall Data Cleaning Observations
**Missing Values**

- There are a large number of missing values in the fields "How_Long_Delayed" which is important to our analysis. 
- There is an extremely large number of missing values in the "Incident_number" field but this is not incidental to our analysis and cannot be filled in without additional information.

**Error in values**

- "How_Long_Delayed" contains string values such as "MINS" or "mins" and a range of values, which needs to be changed to single integer value for our analysis.

**Error in Datatypes**

- "How_Long_Delayed" is a string datatype, should be converted to integer type.

**Error in field names**
- The column name "Boro" should be renamed to "Borough".


### Next Steps: Data Cleaning Steps

**Error in values**

- Extract the first integer value (lowest delay time) in the column "How_Long_Delayed"


**Missing Values**

- Impute the missing values in the field "How_Long_Delayed" with the mean value.


**Error in Datatypes**

- Convert "How_Long_Delayed" to int datatype.



**Error in field names**

- Rename the column "Boro" to "Borough".

In [47]:
# Lets first start by creating a copy of our dataframe 
# df_clean = df.copy(). We will use this copy as our cleaning copy.
# ---
#
Bus_breakdown_clean = Bus_breakdown.copy()
Bus_breakdown_clean.head()

SCHOOL_YEAR  BUSBREAKDOWN_ID  ... BREAKDOWN_OR_RUNNING_LATE SCHOOL_AGE_OR_PREK
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
5   2015-2016          1227543  ...              Running Late         School-Age
6   2015-2016          1227544  ...              Running Late         School-Age

[5 rows x 21 columns]

In [48]:
# Then extracting the lowest delay time in the column how_long_delayed from the string
# 
#Bus_breakdown_clean['how_long_delayed'] = Bus_breakdown_clean['HOW_LONG_DELAYED'].str.extract('(\d+)')
Bus_breakdown_low = Bus_breakdown['HOW_LONG_DELAYED'].min()
Bus_breakdown_low

0

We impute the null values in 'how_long_delayed' column with mean of the column. This will take a couple of steps...

In [63]:
# We first convert our how_long_delayed to float type to allow for imputation
# ---

Bus_breakdown['HOW_LONG_DELAYED'] =Bus_breakdown['HOW_LONG_DELAYED'].astype(float)


In [65]:
# Then later perform our mean imputation 
# ---
#
Bus_breakdown['HOW_LONG_DELAYED'] =Bus_breakdown['HOW_LONG_DELAYED'].fillna(Bus_breakdown['HOW_LONG_DELAYED'].mean())

In [66]:
# Then convert back our how_long_delayed column to integer datatype
# ---
# 
Bus_breakdown['HOW_LONG_DELAYED'] =Bus_breakdown['HOW_LONG_DELAYED'].astype(int)

In [67]:
# Then check for nulls in the column
# ---
# 
Bus_breakdown['HOW_LONG_DELAYED'].isnull()

1         False
2         False
3         False
5         False
6         False
          ...  
281104    False
281105    False
281106    False
281107    False
281109    False
Name: HOW_LONG_DELAYED, Length: 245002, dtype: bool

In [68]:
# Rename Boro column to Borough
# ---
# df.rename(columns={"revenue": "sales"}, inplace=True)
Bus_breakdown.rename(columns = {"BORO":"BOROUGH"}, inplace=True)
Bus_breakdown.columns

Index(['SCHOOL_YEAR', 'BUSBREAKDOWN_ID', 'RUN_TYPE', 'BUS_NO', 'ROUTE_NUMBER',
       'REASON', 'SCHOOLS_SERVICED', 'OCCURRED_ON', 'CREATED_ON', 'BOROUGH',
       'BUS_COMPANY_NAME', 'HOW_LONG_DELAYED', 'NUMBER_OF_STUDENTS_ON_THE_BUS',
       'HAS_CONTRACTOR_NOTIFIED_SCHOOLS', 'HAS_CONTRACTOR_NOTIFIED_PARENTS',
       'HAVE_YOU_ALERTED_OPT', 'INFORMED_ON', 'INCIDENT_NUMBER',
       'LAST_UPDATED_ON', 'BREAKDOWN_OR_RUNNING_LATE', 'SCHOOL_AGE_OR_PREK'],
      dtype='object')

In [69]:
# Lastly we convert all values in our columns to lower case
# for ease of reading
# ---
#df = df.applymap(lambda s:s.lower() if type(s) == str else s)
Bus_breakdown = Bus_breakdown.applymap(lambda s:s.lower() if type(s) == str else s)
#Bus_breakdown = Bus_breakdown.columns.str.lower()
Bus_breakdown 

SCHOOL_YEAR  ...  SCHOOL_AGE_OR_PREK
1        2015-2016  ...          school-age
2        2015-2016  ...               pre-k
3        2015-2016  ...          school-age
5        2015-2016  ...          school-age
6        2015-2016  ...          school-age
...            ...  ...                 ...
281104   2016-2017  ...               pre-k
281105   2016-2017  ...               pre-k
281106   2016-2017  ...               pre-k
281107   2016-2017  ...          school-age
281109   2016-2017  ...          school-age

[245002 rows x 21 columns]

In [70]:
# Check the first 5 record the cleaned dataset
# ---
#
Bus_breakdown.head()

SCHOOL_YEAR  BUSBREAKDOWN_ID  ... BREAKDOWN_OR_RUNNING_LATE SCHOOL_AGE_OR_PREK
1   2015-2016          1227539  ...              running late         school-age
2   2015-2016          1227540  ...              running late              pre-k
3   2015-2016          1227541  ...              running late         school-age
5   2015-2016          1227543  ...              running late         school-age
6   2015-2016          1227544  ...              running late         school-age

[5 rows x 21 columns]


## 5. Solution Implementation

Here we investigate the questions that would help craft our recommendations.

### 5.a) Questions

In [71]:
# 1. Which bus companies that had the highest breakdowns?
# ---
#
Highest_Bus_breakdown = Bus_breakdown.groupby(['BUS_COMPANY_NAME']).count()
# Sort to get bus company with highest breakdowns
#---
# YOUR CODE GOES BELOW
Highest_Bus_breakdown.sort_values (by='BREAKDOWN_OR_RUNNING_LATE', ascending=0).head(1)

SCHOOL_YEAR  ...  SCHOOL_AGE_OR_PREK
BUS_COMPANY_NAME               ...                    
g.v.c., ltd.            19042  ...               19042

[1 rows x 20 columns]

In [72]:
# 2. What were the top 3 reasons for bus delays?
# ---
#
Bus_breakdown_delay = Bus_breakdown.groupby(['REASON']).count()[['HOW_LONG_DELAYED']].reset_index()

# Sort to get the most frequent reason
# ---
# YOUR CODE GOES BELOW
Bus_breakdown_delay.sort_values(by='HOW_LONG_DELAYED', ascending = 0).head(3)

REASON  HOW_LONG_DELAYED
3       heavy traffic            170062
6               other             33122
5  mechanical problem             12597

In [73]:
# 3. How many students were in the buses when they broke down? 
# ---
#
Bus_breakdown.groupby(['NUMBER_OF_STUDENTS_ON_THE_BUS']).count()[['BUSBREAKDOWN_ID']].reset_index()

NUMBER_OF_STUDENTS_ON_THE_BUS  BUSBREAKDOWN_ID
0                                0           142449
1                                1            14069
2                                2            15427
3                                3            13863
4                                4            11417
..                             ...              ...
209                           6109                1
210                           6219                1
211                           8547                1
212                           9007                1
213                           9052                1

[214 rows x 2 columns]

In [74]:
# 4. Which were most frequent reasons for bus breakdowns?
# ---
#
breakdown_reasons = Bus_breakdown[Bus_breakdown.BREAKDOWN_OR_RUNNING_LATE == 'breakdown'].groupby(['REASON']).count()
breakdown_reasons
# Sort to get most frequent reasons
# ---
# YOUR CODE GOES BELOW
breakdown_reasons.sort_values(by='BUSBREAKDOWN_ID', ascending = 0)

SCHOOL_YEAR  ...  SCHOOL_AGE_OR_PREK
REASON                           ...                    
mechanical problem            7  ...                   7
flat tire                     3  ...                   3
other                         1  ...                   1

[3 rows x 20 columns]

In [75]:
# 5. What were the most frequent reasons for the bus running late? 
# ---
#
reasons = Bus_breakdown.groupby(['REASON','HOW_LONG_DELAYED']).count()
reasons
# Get the records with running late reasons and sort to get most frequent reasons
# ---
# YOUR CODE GOES BELOW
reasons.sort_values (by='REASON', ascending=0)

SCHOOL_YEAR  ...  SCHOOL_AGE_OR_PREK
REASON      HOW_LONG_DELAYED               ...                    
won`t start 5594                        1  ...                   1
            32                          1  ...                   1
            0                         154  ...                 154
            1                         178  ...                 178
            2                           1  ...                   1
...                                   ...  ...                 ...
accident    120                         6  ...                   6
            454                         2  ...                   2
            1003                        1  ...                   1
            2308                        1  ...                   1
            0                          32  ...                  32

[491 rows x 19 columns]

In [76]:
# 6. What was the average delay time for each reason type?
# ---
# 
avg_delay = Bus_breakdown.groupby('REASON').mean().reset_index()
avg_delay
# Get the records with reasons and how long on average a delay took then sort 
# ---
# YOUR CODE GOES BELOW
#
avg_delay.sort_values (by='HOW_LONG_DELAYED', ascending=0)

REASON  ...  NUMBER_OF_STUDENTS_ON_THE_BUS
4  late return from field trip  ...                       4.539642
0                     accident  ...                       6.907053
5           mechanical problem  ...                       0.991109
9                  won`t start  ...                       0.568676
7                  problem run  ...                       6.487651
2                    flat tire  ...                       2.747683
8           weather conditions  ...                       2.409736
6                        other  ...                       2.694040
3                heavy traffic  ...                       3.946561
1            delayed by school  ...                       5.994907

[10 rows x 4 columns]

### 5.b) Recommendations

From the above analysis, below are our recommendations:

1. g.v.c. ltd had the highest bus breakdowns and thus should service their buses to avoid breakdowns.
2. Heavy traffic, other and mechanical problems were the major reasons for bus delays hence bus companies should monitor and take routes with lesser traffic and service their buses to avoid delays caused by mechanical problems.
3. In most of the instances the buses were occupied whenever they had breakdowns and hence bus companies should make arrangements for backup buses to pick students and drop them whenever breakdowns happen.
4. Mechanical problems,flat tire and other were the most frequent reasons for bus breakdowns thus bus companies should ensure that their buses are always in good service and have well treaded tires
5. Buses that won't start were the main reasons for bus delays hence bus companies should ensure proper service is done around that area.


## 6. Challenging your Solution

During this step, we review our solution and implement  approaches that could potentially provide a better outcome. In our case, we could propose the following question that wasn't answered in our solution because it couldn't have greatly contributed to our recommendation.

In [78]:
# Which boroughs experienced the most breakdowns?
# ---  
#
breakdowns = Bus_breakdown.groupby(['BOROUGH']).count()
breakdowns
# Sort to get bus company with highest breakdowns
# ---
# YOUR CODE GOES BELOW
#
breakdowns.sort_values(by='BUSBREAKDOWN_ID', ascending=0)

SCHOOL_YEAR  ...  SCHOOL_AGE_OR_PREK
BOROUGH                       ...                    
manhattan              63823  ...               63823
bronx                  60908  ...               60908
brooklyn               53324  ...               53324
queens                 28350  ...               28350
staten island          13257  ...               13257
westchester             6906  ...                6906
nassau county           3409  ...                3409
new jersey              1423  ...                1423
rockland county          821  ...                 821
all boroughs             402  ...                 402
connecticut              200  ...                 200

[11 rows x 20 columns]

Our observations:

1. Manhattan had the most breakdowns  

How does this observation tie to our solution?



## 7. Follow up questions

During this step, you rethink and propose other ways that you can improve your solution.

### a). Did we have the right data?

Yes

### b). Do we need other data to answer our question?

You can look into the questions you brainstormed that you weren't taken into account during analysis due to a lack of data. Were those questions important to have been left out  of your analysis?

No

### c). Did we have the right question?

Were there any other questions that we needed to have answered?

Yes